extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -0,0 +1,208 @@
+                return ConversionResult.Unhandled();
+            }
+
+            try
+            {
+                var cosmosAttribute = GetBindingDataContent(modelBindingData);
+                object result = await ToTargetTypeAsync(context.TargetType, cosmosAttribute);
+
+                if (result is not null)
+                {
+                    return ConversionResult.Success(result);
+                }
+            }
+            catch (Exception ex)
+            {
+                return ConversionResult.Failed(ex);
+            }
+
+            return ConversionResult.Unhandled();
+        }
+
+        private bool IsCosmosExtension(ModelBindingData bindingData)
+        {
+            if (bindingData?.Source is not Constants.CosmosExtensionName)
+            {
+                _logger.LogTrace("Source '{source}' is not supported by {converter}", bindingData?.Source, nameof(CosmosDBConverter));
+                return false;
+            }
+
+            return true;
+        }
+
+        private CosmosDBInputAttribute GetBindingDataContent(ModelBindingData bindingData)
+        {
+            return bindingData?.ContentType switch
+            {
+                Constants.JsonContentType => bindingData.Content.ToObjectFromJson<CosmosDBInputAttribute>(),
+                _ => throw new NotSupportedException($"Unexpected content-type. Currently only '{Constants.JsonContentType}' is supported.")
+            };
+        }
+
+        private async Task<object> ToTargetTypeAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute) => targetType switch
+        {
+            Type _ when targetType == typeof(CosmosClient) => CreateCosmosClient<CosmosClient>(cosmosAttribute),
+            Type _ when targetType == typeof(Database) => CreateCosmosClient<Database>(cosmosAttribute),
+            Type _ when targetType == typeof(Container) => CreateCosmosClient<Container>(cosmosAttribute),
+            _ => await CreateTargetObjectAsync(targetType, cosmosAttribute)
+        };
+
+        private async Task<object> CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute)
+        {
+            MethodInfo createPOCOMethod;
+
+            if (targetType.GenericTypeArguments.Any())
+            {
+                targetType = targetType.GenericTypeArguments.FirstOrDefault();
+
+                createPOCOMethod = GetType()
+                                    .GetMethod(nameof(CreatePOCOCollectionAsync), BindingFlags.Instance | BindingFlags.NonPublic)
+                                    .MakeGenericMethod(targetType);
+            }
+            else
+            {
+                createPOCOMethod = GetType()
+                                    .GetMethod(nameof(CreatePOCOAsync), BindingFlags.Instance | BindingFlags.NonPublic)
+                                    .MakeGenericMethod(targetType);
+            }
+
+
+            var container = CreateCosmosClient<Container>(cosmosAttribute) as Container;
```

Comment:
Is this creating a new cosmos client each time? In WebJobs, we maintain a cache based on (i think) the connection string and the preferredlocations string so we re-use clients whenever possible...

https://github.com/Azure/azure-webjobs-sdk-extensions/blob/8b66ca0511cc101794e6492759729db0fbb0029d/src/WebJobs.Extensions.CosmosDB/Config/CosmosDBExtensionConfigProvider.cs#L117-L124

Reply:
Oh nice! Thats even better, will implement a cache here as well